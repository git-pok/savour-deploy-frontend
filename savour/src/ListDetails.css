.ListDetails {
  display: block;
  width: 100%;
  height: 100%;
}

.ListDetails-div {
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex-wrap: wrap;
  width: 70%;
  height: 100%;
  margin: 0px auto 140px auto;
  padding: 30px 30px;
  box-sizing: content-box;
  border-radius: 10px;
  /* background-color: rgba(186, 183, 108, 0.371); */
  background-color: rgb(216, 216, 216);
}

.ListDetails-split-ul-div {
  float: left;
  width: 50%;
  margin: 0px auto;
}

.ListDetails-ul {
  list-style: circle;
  width: 100%;
  padding-left: 13px;
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  align-items: center;
}

.ListDetails-icons {
  margin: 10px 10px;
  font-size: 30px;
  transition: .5s ease-in-out background-color;
}

.ListDetails-li-icon-div {
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: center;
}

.ListDetails-li-icon {
  margin: 0px 10px;
  font-size: 20px;
  transition: .5s ease-in-out background-color;
}

.hover-green:hover {
  color: green;
}

.hover-red:hover {
  color: red;
}

.ListDetails-icons-center-div {
  /* float: none; */
  display: block;
  width: 100%;
  margin: 0px auto;
}

.ListDetails-icon-center {
  margin: 10px 10px;
  font-size: 30px;
  transition: .5s ease-in-out background-color;
}

.ListDetails-ul li {
  width: 400px;
  text-align: left;
  margin: 10px 10px;
}

.ListDetails-h1 {
  width: 100%;
  font-size: 60px;
  margin-bottom: 70px;
  /* color: rgb(255, 255, 255); */
}

.ListDetails-msg {
  display: block;
  width: 100%;
}

.ListDetails-msg div {
  display: block;
  margin: 0px auto;
}

.ListDetails-split-div {
  display: inline-block;
  width: 70%;
  height: 100%;
  margin: 0px auto 140px auto;
  padding: 30px 30px;
  box-sizing: content-box;
  border-radius: 10px;
  background-color: rgb(216, 216, 216);
}

.ListDetails-split-content-div {
  display: block;
  float: left;
  width: 50%;
  margin: 0px auto;
}

.ListDetails-subtitle {
  text-align: center;
  display: block;
  font-size: 26px;
  margin: 0px auto -10px auto;
}

.ListDetails-p {
  display: block;
  width: 90%;
  margin: 10px auto 20px auto;
  text-align: left;
}

.ListDetails-short-text {
  display: block;
  width: 90%;
  margin: 10px auto 20px auto;
  text-align: center;
}

.ListDetails-short-text:link {
  display: block;
  width: 90%;
  margin: 10px auto 20px auto;
  text-align: center;
  text-decoration: none;
}


@media only screen and (max-width: 1302px) and (min-width: 400px) {
  .ListDetails-ul {
    align-items: center;
  }

  .ListDetails-ul li {
    width: 300px;
  }
  
  .ListDetails-div {
    flex-direction: column;
    align-items: center;
    flex-wrap: nowrap;
  }
  .ListDetails-split-ul-div {
    float: none;
    width: 100%;
    margin-bottom: 70px;
  }

  .ListDetails-split-content-div {
    display: block;
    float: none;
    width: 100%;
  }
}
