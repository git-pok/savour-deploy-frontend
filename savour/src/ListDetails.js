import { useContext, useState, useEffect } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import UserContext from './context/UserContext.js';
import SavourApi from './models/SavourApi.js';
import RecipeContainer from './RecipeContainer.js';
import Message from './Message.js';
import AddToListForm from './AddToListForm.js';
import DeleteFromRecipelistForm from './DeleteFromRecipelistForm.js';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {
          faSquarePlus, faSquareMinus, faRectangleXmark
        } from '@fortawesome/free-solid-svg-icons';
import useToggleState from './hooks/useToggleState.js';
import './ListDetails.css';

/**
 * ListDetails
 * Props: urlEndpt, recipelist
 * Renders html for shopping or recipe list details.
*/
const ListDetails = ({urlEndpt, recipelist=false}) => {
  const { usrData, setUsrData } = useContext(UserContext);
  const headers = { _token: `Bearer ${usrData.token}`};
  // Destructure list id params from route.
  const { id: listId } = useParams();
  const history = useHistory();
  const listUrl = `/users/${usrData.userId}/${urlEndpt}/${listId}`;

  const [ activateGetLists, setActivateGetLists ] = useToggleState(false);
  const [ listData, setListData ] = useState(null);
  const [ isIngrAddBtn, setIsIngrAddBtn ] = useToggleState(false);
  const [ isIngrRmvBtn, setIsIngrRmvBtn ] = useToggleState(false);
  const [ isListDltBtn, setIsListDltBtn ] = useToggleState(false);
  const [ isAddActionCmplt, setIsAddActionCmplt ] = useToggleState(false);
  const [ isRmvActionCmplt, setIsRmvActionCmplt ] = useToggleState(false);
  const [ isLstDltActionCmplt, setIsLstDltActionCmplt ] = useToggleState(false);
  const [ isAddSucc, setIsAddSucc ] = useToggleState(false);
  const [ isRmvSucc, setIsRmvSucc ] = useToggleState(false);
  const [ formErrMsg, setFormErrMsg ] = useState(null);
  const [ succMsg, setSuccMsg ] = useState(null);
  const [ shoplistItems, setShoplistItems ] = useState(null);
  const [ updateListMsg, setUpdateListMsg ] = useState(null);

  const addOrRmvItem = (itemsObj, add=true) => {
    if (add) setIsIngrAddBtn();
    else setIsIngrRmvBtn();
    setShoplistItems(() => ({
      ...itemsObj
    }));
  }

  const deleteList = () => {
    setIsListDltBtn();
  }

  useEffect(() => {
    const getLists = async () => {
      try {
        const listReq = await SavourApi.request("get", listUrl, {}, {}, headers);
        setListData(() => listReq.data);
        if (activateGetLists) setTimeout(setActivateGetLists, 3000);
      } catch(err) {
        console.log(`Error!\n${err}`);
      }
    }

    const editList = async () => {
      try {
        const url = `${listUrl}/items`;
        const method = isIngrRmvBtn ? "delete" : "post";
        const listReq = await SavourApi.request(method, url, shoplistItems, {}, headers);
        const msg = isIngrAddBtn ? "Added item to list!" : "Removed item from list!";
        setSuccMsg(() => msg);
        isIngrAddBtn ? setIsAddActionCmplt() : setIsRmvActionCmplt();
        isIngrAddBtn ? setIsAddSucc() : setIsRmvSucc();
        isIngrAddBtn ? setTimeout(setIsAddActionCmplt, 3000) : setTimeout(setIsRmvActionCmplt, 3000);
        isIngrAddBtn ? setTimeout(setIsAddSucc, 3000) : setTimeout(setIsRmvSucc, 3000);
        setTimeout(() => setSuccMsg(null), 3000);
        setTimeout(() => setShoplistItems(null), 3000);
        if (isIngrAddBtn) setTimeout(setIsIngrAddBtn, 3000);
        if (isIngrRmvBtn) setTimeout(setIsIngrRmvBtn, 3000);
      } catch (err) {
        const error = err.response.data.error.message;
        setFormErrMsg(() => error || "Error");
        isIngrAddBtn ? setIsAddActionCmplt() : setIsRmvActionCmplt();
        isIngrAddBtn ? setTimeout(setIsAddActionCmplt, 3000) : setTimeout(setIsRmvActionCmplt, 3000);
        setTimeout(() => setFormErrMsg(null), 3000);
        isIngrAddBtn ? setTimeout(setIsIngrAddBtn, 3000) : setTimeout(setIsIngrRmvBtn, 3000);
      }
    }

    const deleteList = async () => {
      try {
        const listReq = await SavourApi.request("delete", listUrl, {}, {}, headers);
        const msgListName = recipelist ? "recipelists" : "shoppinglists";
        const msg = `Deleted list! Redirecting to ${msgListName}!`;
        setSuccMsg(() => msg);
        setIsLstDltActionCmplt();
        setIsRmvSucc();
        setTimeout(setIsLstDltActionCmplt, 3000);
        setTimeout(setIsRmvSucc, 3000);
        setTimeout(() => setSuccMsg(null), 3000);
        setTimeout(setIsListDltBtn, 3000);
        setTimeout(() => history.push(`/${urlEndpt}`), 1500);
      } catch (err) {
        const error = err.response.data.error.message;
        setFormErrMsg(() => error || "Error");
        setTimeout(setIsLstDltActionCmplt, 3000);
        setTimeout(setIsListDltBtn, 3000);
        setTimeout(() => setFormErrMsg(null), 3000);
      }
    }

    if (listUrl) getLists();
    if (isIngrAddBtn || isIngrRmvBtn) editList();
    if (isListDltBtn) deleteList();
  }, [listUrl, activateGetLists, isIngrAddBtn, isIngrRmvBtn, isListDltBtn])

  return (
    <>
    { !recipelist ?
        <h1 className="ListDetails-h1">{`${listData && listData.list_name} List`}</h1>
      :
        <h1 className="ListDetails-h1">{`${listData && listData.list_name} for ${listData && listData.occasion}  Recipes`}</h1>
    }
    { listData && recipelist && listData.recipes.length ?
    <div className="ListDetails">
      <div className="ListDetails-div">
        <RecipeContainer
          showHide={true}
          recipeArray={listData.recipes} />
        <div className="ListDetails-icons-center-div">
          <h2 className="ListDetails-subtitle">Delete Recipelist</h2>
          <FontAwesomeIcon
            onClick={deleteList}
            className="ListDetails-icon-center hover-red"
            icon={faRectangleXmark} />
            <div className="ListDetails-msg">
              {
                isLstDltActionCmplt &&
                <Message msgObj={
                  {
                    class: isRmvSucc ? "success" : "fail",
                    msg: isRmvSucc ? succMsg : formErrMsg
                  }
                } />
              }
            </div>
        </div>
      </div>
    </div>
    :
      null
    }
    { listData && recipelist && !listData.recipes.length ?
      <div className="ListDetails">
        <div className="ListDetails-div">
          <p>No recipes</p>
          <div className="ListDetails-icons-center-div">
            <h2 className="ListDetails-subtitle">Delete Recipelist</h2>
            <FontAwesomeIcon
              onClick={deleteList}
              className="ListDetails-icon-center hover-red"
              icon={faRectangleXmark} />
            <div className="ListDetails-msg">
              {
                isLstDltActionCmplt &&
                <Message msgObj={
                  {
                    class: isRmvSucc ? "success" : "fail",
                    msg: isRmvSucc ? succMsg : formErrMsg
                  }
                } />
              }
            </div>
          </div>
        </div>
      </div>
    :
      null
    }
    { listData && recipelist &&
      <div className="ListDetails">
        <h1 className="ListDetails-h1">Add Recipes to List</h1>
        <div className="ListDetails-div">
          <AddToListForm recipelist={true} setToggleState={setActivateGetLists} />
        </div>
      </div>
    }
    {/* DELETE RECIPE FROM RECIPELISTS */}
    { listData && recipelist &&
      <div className="ListDetails">
        <h1 className="ListDetails-h1">Delete Recipe from List</h1>
        <div className="ListDetails-div">
          <DeleteFromRecipelistForm
            recipelistId={listId}
            setState={setListData}
            list={listData} />
        </div>
      </div>
    }
    {/* DELETE RECIPELIST FROM RECIPELISTS END */}
    {/* SHOPPINGLIST DETAILS */}
    { listData && !recipelist ?
        <div className="ListDetails-split-div">
          <div className="ListDetails-split-ul-div">
            <h2 className="ListDetails-subtitle">Recipe's Ingredients</h2>
            <ul className="ListDetails-ul">
              { listData.recipes_ingredients.map(ingrd => (
                <div key={`div-${ingrd.id}`} className="ListDetails-li-icon-div">
                  <li>{ingrd.qty}{ingrd.unit !== "no unit" ? ingrd.unit : null} {ingrd.ingredient}</li>
                  <FontAwesomeIcon
                    className="ListDetails-li-icon hover-green"
                    onClick={() => addOrRmvItem({qty: ingrd.qty, unit_id: ingrd.unit_id, ingredient_id: ingrd.ingredient_id})}
                    icon={faSquarePlus} />
                </div>
                ))
              }
            </ul>
            <div className="ListDetails-msg">
              {
                isAddActionCmplt &&
                  <Message msgObj={
                    {
                      class: isAddSucc ? "success" : "fail",
                      msg: isAddSucc ? succMsg : formErrMsg
                    }
                  } />
              }
            </div>
          </div>

          <div className="ListDetails-split-content-div">
            <h2 className="ListDetails-subtitle">List Name</h2>
            <p className="ListDetails-short-text">{listData.list_name}</p>

            <h2 className="ListDetails-subtitle">Recipe Name</h2>
            <p className="ListDetails-short-text">{listData.recipe_name}</p>

            <h2 className="ListDetails-subtitle">Recipe Author</h2>
            <p className="ListDetails-short-text">{listData.recipe_author}</p>
            <h2 className="ListDetails-subtitle">Delete Shoppinglist</h2>
            <FontAwesomeIcon
              className="ListDetails-icons hover-red"
              onClick={deleteList}
              icon={faRectangleXmark} />
            <div className="ListDetails-msg">
              {
                isLstDltActionCmplt &&
                  <Message msgObj={
                    {
                      class: isRmvSucc ? "success" : "fail",
                      msg: isRmvSucc ? succMsg : formErrMsg
                    }
                  } />
              }
            </div>
          </div> 
        </div>
      :
        null
    }
    { listData && !recipelist &&
      <h1 className="ListDetails-h1">Shoppinglist Items</h1>
    }

    { listData && !recipelist ?
      <div className="ListDetails">
        <div className="ListDetails-div">
          {
            !listData.list_items.length ?
              <h1>Add Items</h1>
            :
              <ul className="ListDetails-ul">
                { listData.list_items.map(item => (
                  <div key={`div-${item.id}`} className="ListDetails-li-icon-div">
                    <li key={`li-${item.id}`}>{item.qty || null}{item.unit !== "no unit" ? item.unit : null} {item.ingredient}</li>
                    <FontAwesomeIcon
                      className="ListDetails-li-icon hover-red"
                      onClick={() => addOrRmvItem({item_id: item.id}, false)}
                      icon={faSquareMinus} />
                  </div>
                ))}
              </ul>
          }
          <div className="ListDetails-msg">
            { isRmvActionCmplt &&
              <Message msgObj={
                {
                  class: isRmvSucc ? "success" : "fail",
                  msg: isRmvSucc ? succMsg : formErrMsg
                }
              } />
            }
          </div>
        </div>
      </div>
    :
      null
    }
    </>
  );
}

export default ListDetails;