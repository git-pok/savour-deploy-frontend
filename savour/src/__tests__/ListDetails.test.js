import ListDetails from '../ListDetails.js';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import UserContext from '../context/UserContext.js';

const usrData = {
  userUsername: "fvin",
  userId: 12
};

test("renders ListDetails without crashing", () => {
  render(
    <MemoryRouter>
      <UserContext.Provider value={{usrData}}>
        <ListDetails
          urlEndpt="/recipes" />
      </UserContext.Provider>
    </MemoryRouter>
  );
});

test("renders snapshot without crashing", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <UserContext.Provider value={{usrData}}>
        <ListDetails
          urlEndpt="/recipes" />
      </UserContext.Provider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});