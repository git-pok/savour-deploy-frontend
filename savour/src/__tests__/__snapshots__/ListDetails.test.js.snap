// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot without crashing 1`] = `
<DocumentFragment>
  <h1
    class="ListDetails-h1"
  >
    null List
  </h1>
</DocumentFragment>
`;
